### Pair Programming - ANOVA

Hasta ahora habéis estado evaluando las características de vuestro set de datos y habéis hecho una gran exploración, es el momento de hacer vuestro primer ANOVA! En el ejercicio de hoy tendréis que hacer un ANOVA con vuestro datos y hacer una interpretación de los resultados.

NOTA Puede que vuestros datos no se ajusten o no cumplan todas las asunciones, no pasa nada, haced el ANOVA e interpretad los resultados. En próximas lecciones aprenderemos que podemos hacer cuando nos encontramos en esta situación.

In [14]:
import numpy as np
import pandas as pd
import random 

import matplotlib.pyplot as plt
import seaborn as sns

import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.multivariate.manova import MANOVA
from sklearn.preprocessing import StandardScaler

import statsmodels.api as sm
from statsmodels.formula.api import ols

plt.rcParams["figure.figsize"] = (10,8) 

In [15]:
df = pd.read_csv("data/adult.data_limpio.csv", index_col = 0)

In [16]:
df.head(2)

work_class  final_weight   education  education_yrs  \
39                                                               
50   Self-emp-not-inc         83311   Bachelors             13   
38            Private        215646     HS-grad              9   

         marital_status          occupation    relationship ethnicity gender  \
39                                                                             
50   Married-civ-spouse     Exec-managerial         Husband     White   Male   
38             Divorced   Handlers-cleaners   Not-in-family     White   Male   

    capital_gain  capital_lost  hours_week         country  salary census  
39                                                                         
50             0             0          13   United-States   <=50K   Bajo  
38             0             0          40   United-States   <=50K   Bajo

In [17]:
anova = ols('final_weight ~ work_class  + education + education_yrs + marital_status + occupation + relationship + ethnicity + gender + capital_gain + capital_lost + hours_week + country + salary + census' , data=df).fit()
sm.stats.anova_lm(anova)

df        sum_sq       mean_sq            F  \
work_class          8.0  1.003162e+12  1.253952e+11    12.705783   
education          15.0  1.503840e+12  1.002560e+11    10.158531   
marital_status      6.0  8.366524e+11  1.394421e+11    14.129096   
occupation         14.0  7.278147e+11  5.198676e+10     5.267606   
relationship        5.0  1.507296e+11  3.014592e+10     3.054563   
ethnicity           4.0  6.669374e+12  1.667343e+12   168.945106   
gender              1.0  6.698914e+11  6.698914e+11    67.877359   
country            41.0  8.188998e+12  1.997317e+11    20.237994   
salary              1.0  9.944195e+10  9.944195e+10    10.076046   
census              1.0  2.243753e+13  2.243753e+13  2273.503608   
education_yrs       1.0  1.422502e+09  1.422502e+09     0.144136   
capital_gain        1.0  8.264308e+09  8.264308e+09     0.837389   
capital_lost        1.0  8.493063e+09  8.493063e+09     0.860567   
hours_week          1.0  6.806454e+10  6.806454e+10     6.896702   
Residual        32461.0  3.203623e+14  9.869143e+09          NaN   

                       PR(>F)  
work_class       2.109971e-18  
education        9.434647e-25  
marital_status   3.865187e-16  
occupation       4.135104e-10  
relationship     9.269151e-03  
ethnicity       1.922569e-143  
gender           1.804504e-16  
country         1.635877e-145  
salary           1.503495e-03  
census           0.000000e+00  
education_yrs    7.042056e-01  
capital_gain     3.601515e-01  
capital_lost     3.535869e-01  
hours_week       8.639543e-03  
Residual                  NaN

**Interpretación de la ANOVA:**

Es importante recalcar previo a la interpretación que los datos con los que trabajamos no son normales.
- **df:** Las variables categóricas tienen más rango de libertad, mientras que las numéricas sólo uno. Esto es porque las numéricas sólo tienen el rango de los números naturales.
- **sum_sq:** Esta es la variación atribuida al error. Podemos observar que, en este caso, es mínima.
- **mean_sq:** Este es el producto de la dos columnas anteriores.
- **F:** ANOVA supone que las medias de todos los grupos son iguales, calcula cuánto se desvían las medias iguales de la suposición. Una puntuación mayor significa que hay mayor diferencia entre las medias, por lo tanto, las que tienen valores más próximos entre sí van a tener medias más similares.
- **PR(>F):** Aquí podemos observar que el p-valor es en todos los casos inferior a 0,05, con lo cual, debemos rechazar la hipótesis nula y concluimos que estas variables sí influyen en nuestra variable predictora.

In [18]:
anova.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:           final_weight   R-squared:                       0.117
Model:                            OLS   Adj. R-squared:                  0.114
Method:                 Least Squares   F-statistic:                     43.81
Date:                Mon, 23 Jan 2023   Prob (F-statistic):               0.00
Time:                        10:30:26   Log-Likelihood:            -4.2080e+05
No. Observations:               32560   AIC:                         8.418e+05
Df Residuals:                   32461   BIC:                         8.426e+05
Df Model:                          98                                         
Covariance Type:            nonrobust                                         
============================================================================================================
                                               coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------
Intercept                                 8.025e+05   1.81e+04     44.271      0.000    7.67e+05    8.38e+05
work_class[T. Federal-gov]               -3546.4328   3968.538     -0.894      0.372   -1.13e+04    4232.049
work_class[T. Local-gov]                  2763.8696   3411.569      0.810      0.418   -3922.932    9450.671
work_class[T. Never-worked]               2.651e+04   3.76e+04      0.704      0.481   -4.73e+04       1e+05
work_class[T. Private]                    5126.6299   2718.461      1.886      0.059    -201.655    1.05e+04
work_class[T. Self-emp-inc]              -6796.9065   3975.090     -1.710      0.087   -1.46e+04     994.418
work_class[T. Self-emp-not-inc]          -3479.4005   3295.188     -1.056      0.291   -9938.091    2979.290
work_class[T. State-gov]                 -1687.2137   3757.164     -0.449      0.653   -9051.394    5676.967
work_class[T. Without-pay]                2041.2730   2.54e+04      0.080      0.936   -4.78e+04    5.19e+04
education[T. 11th]                       -1.356e+04   4207.093     -3.223      0.001   -2.18e+04   -5313.387
education[T. 12th]                       -2.586e+04   5556.213     -4.654      0.000   -3.68e+04    -1.5e+04
education[T. 1st-4th]                     3.964e+04   8828.418      4.490      0.000    2.23e+04    5.69e+04
education[T. 5th-6th]                     2.125e+04   6997.647      3.036      0.002    7531.677     3.5e+04
education[T. 7th-8th]                     9858.0903   5439.919      1.812      0.070    -804.353    2.05e+04
education[T. 9th]                         4387.1102   5621.711      0.780      0.435   -6631.652    1.54e+04
education[T. Assoc-acdm]                  -6.26e+04   3914.679    -15.991      0.000   -7.03e+04   -5.49e+04
education[T. Assoc-voc]                  -6.132e+04   3686.143    -16.636      0.000   -6.85e+04   -5.41e+04
education[T. Bachelors]                  -7.771e+04   3073.271    -25.287      0.000   -8.37e+04   -7.17e+04
education[T. Doctorate]                  -1.075e+05   5133.910    -20.941      0.000   -1.18e+05   -9.74e+04
education[T. HS-grad]                     -3.57e+04   2914.569    -12.250      0.000   -4.14e+04      -3e+04
education[T. Masters]                    -9.429e+04   3639.926    -25.904      0.000   -1.01e+05   -8.72e+04
education[T. Preschool]                   5.295e+04   1.41e+04      3.749      0.000    2.53e+04    8.06e+04
education[T. Prof-school]                -9.734e+04   4732.540    -20.569      0.000   -1.07e+05   -8.81e+04
education[T. Some-college]               -4.733e+04   2920.024    -16.208      0.000   -5.31e+04   -4.16e+04
marital_status[T. Married-AF-spouse]      1.666e+04   2.17e+04      0.766      0.443   -2.59e+04    5.93e+04
marital_status[T. Married-civ-s

R-cuadrado = Variación explicada / variación total
El R-cuadrado siempre está entre 0 y 100%:
0% indica que el modelo no explica ninguna porción de la variabilidad de los datos de respuesta en torno a su media.
100% indica que el modelo explica toda la variabilidad de los datos de respuesta en torno a su media.

**Interpretación del summary**

- **R-square:** Siempre se encuentra entre el 0 y el 100%, el nuestro está en 0.117, por lo tanto, esto nos quiere decir que la variables presidctoras explican en un 11,7% nuestra variable respuesta.
- **coef:** Estos datos nos indican cuánto son afectadas las variables por las que se relacionan con ellas. Vemos que las afectaciones son mínimas, de muy poquitas milésimas, tanto en positivo como en negativo.
- **std err:** Esta columna nos está diciendo que tenemos un gran margen de error en la estimación del coeficiente, seguramente debido a la cantidad de ouliers.
- **t:** Es el resultado de dividir los dos valores anteriores. Con esto estamos intentando averiguar en qué medida la afectación de los datos entre sí influye en el margen de error, es decir, si las variables se influyen también cuando no tienen correlación (or something???).